flow.py

```diff
@@ -57,23 +66,120 @@ def dest(self):
+            else:
+                self.create_packet(self.num_packets_sent)
+                self.num_packets_sent += 1
+
+    def create_packet(self, packet_id):
+        """Method creates packet and then adds them to event queue to be sent
+        to the host, and adds a timing event to ensure that they are resent if
+        unacknowledged.
+
+        Args:
+            packet_id(int): Unique id identifying the packet
+
+        """
+        new_packet = DataPacket(packet_id, self.src.node_id,
+            self.dest.node_id, self.flow_id)
+
+        print("Flow", self.flow_id, ": made data packet", new_packet.packet_id)
+
+        self.unacknowledged_packets.add(new_packet.packet_id)
+
+        # Adding events to queues
+        event1 = lambda: self.src.send_packet(new_packet)
+        self.ns.add_event(event1)
+        event2 = lambda: self.time_out(new_packet.packet_id)
+        self.ns.add_event(event2, delay=ACK_DELAY)
+
+        print("Flow", self.flow_id, ": sent data packet", new_packet.packet_id)
+
+    def make_acknowledgement_packet(self, packet_id, src, dest, packet_size, acknowledge_id):
+        """Method makes the AcknowledgementPacket and triggers the send_packet
+        method for the host if applicable
 
         Args:
-            id (int): The unique id identifying the packet
+            packet_id (int): Unique id identifying the packet
             src (Node): The packet's source node
             dest (Node): The packet's destination node
-            data (any datatype): The data in the packet
+            packet_size (float): The packet's size in bits
+            flow_id (string): Unique id indicating flow
+
+        """
+        print("Flow ", self.flow_id, ": made acknowledgement packet", packet_id)
+        new_packet = AcknowledgementPacket(ns, packet_id, src, dest,
+            self.flow_id)
+
+        event = lambda: self.src.send_packet(new_packet)
+        self.ns.add_event(event)
+
+    def receive_packet(self, packet):
+        """Method receives a given packet.  If it's a data packet, send an
+        acknowledgement packet.  If it's an acknowledgement packet, update_flow
+
+        args:
+            packet (Packet): packet object being received
+
         """
-        pass
+        print("Flow", self.flow_id, ": received data packet", packet.packet_id)
+        if isinstance(packet, DataPacket):
+            self.acknowledge(packet)
+        elif isinstance(packet, AcknowledgementPacket):
+            self.update_flow(packet)
 
     def acknowledge(self, packet):
         """Method that triggers the send_packet function for the host if
         applicable by sending the acknowledgement packet
 
         Args:
             packet (Packet): The packet attempting to be acknowledged
+
         """
-        pass
+        print("Flow ", self.flow_id, ": sending acknowledgement packet",
+            packet.packet_id)
+        a_packet = make_acknowledgement_packet(packet.packet_id, packet.src,
```

Comment:
Shouldn't src and dest be switched here since the previous destination is now the source and vice verse